aio_pika/channel.py

```diff
@@ -56,158 +55,180 @@ def __init__(
 
     @property
     def done_callbacks(self) -> CallbackCollection:
-        return self._done_callbacks
+        return self.close_callbacks
 
     @property
-    def return_callbacks(self) -> CallbackCollection:
-        return self._return_callbacks
+    def is_opened(self) -> bool:
+        return hasattr(self, "_channel")
 
     @property
-    def is_closed(self):
-        if not self._channel:
+    def is_closed(self) -> bool:
+        if not self.is_opened or self._is_closed_by_user:
             return True
-        return self.channel.is_closed
-
-    async def close(self, exc=None):
-        if not self._channel:
-            log.warning("Channel already closed")
-            return
+        return self._channel.is_closed
 
-        if self.channel.is_closed:
+    @task
+    async def close(self, exc: aiormq.abc.ExceptionType = None) -> None:
+        if not self.is_opened:
+            log.warning("Channel not opened")
             return
 
-        # noinspection PyTypeChecker
-        channel = self._channel  # type: aiormq.Channel
-        self._channel = ()
+        channel: aiormq.abc.AbstractChannel = self._channel
+        del self._channel
+        self._is_closed_by_user = True
         await channel.close()
 
-        self._done_callbacks(exc)
-
     @property
-    def channel(self) -> aiormq.Channel:
-        if self._channel is None:
-            raise RuntimeError("Channel was not opened")
+    def channel(self) -> aiormq.abc.AbstractChannel:
+        if not self.is_opened:
+            raise aiormq.exceptions.ChannelInvalidStateError(
+                "Channel was not opened",
+            )
+
+        if self.is_closed:
+            raise aiormq.exceptions.ChannelInvalidStateError(
+                "Channel has been closed",
+            )
 
         return self._channel
 
     @property
-    def number(self):
-        return self.channel.number if self._channel else None
+    def number(self) -> Optional[int]:
+        return self._channel.number if self.is_opened else None
 
-    def __str__(self):
-        return "{0}".format(self.number or "Not initialized channel")
+    def __str__(self) -> str:
+        return "{}".format(self.number or "Not initialized channel")
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         conn = None
 
-        if self._channel:
+        if self.is_opened:
             conn = self._channel.connection
 
-        return '<%s "%s#%s">' % (self.__class__.__name__, conn, self)
-
-    def __iter__(self):
-        return (yield from self.__await__())
+        return '<%s #%s "%s">' % (self.__class__.__name__, self, conn)
 
-    def __await__(self):
+    def __await__(self) -> Generator[Any, Any, "AbstractChannel"]:
         yield from self.initialize().__await__()
         return self
 
-    async def __aenter__(self):
-        if not self._channel:
+    async def __aenter__(self) -> "AbstractChannel":
+        if not hasattr(self, "_channel"):
```

Comment:
Why not simple None, instead of checking attr?
Some GC feature?

Reply:
Do not want to make it `Optional` and check that every time.

Reply:
But you make it not existing at all and must check's all the time :)

Reply:
C'est La Vie